[Experiment][ConstraintSystem] Introduce `type member` constraint

xedin opened the pull request:
Add a new `baseType[.<associated type>] == memberType` constraint
which replaces use of `DependentMemberType` inside of a
constraint system.

The idea here is two fold:

1. Constraint system simplification (`getFixedTypeRecursive`, `simplifyTypeImpl` no longer need
   to attempt substitutions, holes related to type members are recorded in one place etc.);

2. Removal of stateful tracking related to dependent member type inference
    which paves the way for increment binding computation.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Introduce `type member` constraint

Add a new `baseType[.<associated type>] == memberType` constraint
which replaces use of `DependentMemberType` inside of a
constraint system.

xedin committed:
[ConstraintSystem] Remove `DependentMemberType` simplication logic

Remove obsolete logic and Make sure that neither `getFixedTypeRecursive`
nor `simplifyTypeImpl` receive a `DependentMemberType` for simplication
by adding new assertion.

Add a new `replaceDependentTypes` method which converts `DependentMemberType`
into a type-member constraint plus a type variable (for resulting type).

xedin committed:
[CSBindings] Remove `DependentMemberType` related checks from binding inference

xedin commented:
There is still a lot of work here e.g. new locator element to track information about associated types, diagnostic fixes, but I was able to build stdlib with this changes.

CodaFi commented on lib/Sema/CSBindings.cpp:
```diff
@@ -361,17 +359,6 @@ ConstraintSystem::getPotentialBindingForRelationalConstraint(
-  // If the type we'd be binding to is a dependent member, don't try to
-  // resolve this type variable yet.
-  if (type->is<DependentMemberType>()) {
-    if (!ConstraintSystem::typeVarOccursInType(typeVar, type,
```
Just to confirm: The occurs check is being handled by the normal bindings path?

xedin replied:
Yes, it's verified separately from dependent member types.

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -6832,6 +6790,61 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
+
+  // FIXME: It's kind of weird in general that we have to look
+  // through lvalue, inout and IUO types here
+  auto baseTy = getFixedTypeRecursive(origBaseType, flags, /*wantRValue=*/true)
```
I put this FIXME in a long time ago because I observed DependentMemberTypes where the base was an InOutType of a GenericTypeParamType. This might no longer happen (even without your PR?). Do you want to try removing the call to getWithoutSpecifierType()?

xedin replied:
Sure! I'll try it out!

slavapestov commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -4448,3 +4425,17 @@ void ConstraintSystem::maybeProduceFallbackDiagnostic(
   ctx.Diags.diagnose(target.getLoc(), diag::failed_to_produce_diagnostic);
 }
+
+Type ConstraintSystem::replaceDependentTypes(Type origType) {
```
I feel like this is getting called in the wrong places. We should replace DependentMemberTypes with TypeVariableTypes in the same places where we replace GenericTypeParameterTypes with TypeVariableTypes, ie `openGeneric()` and so on. The constraint system should never see types that contain DependentMemberTypes just like it never sees types that contain GenericTypeParamTypes.

xedin replied:
Tha'ts what I thought as well but in reality after I have added asserts to `getFixedTypeRecursive` and `simplifyTypeImpl` it turned up that conditional conformance requirements has dependent member types and `getSuperclass()` returns type with dependent member types in generic arguments, these types aren't opened. This changes serve more to show the approach so I'm still in search for a right solution how to deal with that...

slavapestov replied:
Can you share examples that demonstrate those problems?

xedin replied:
Sure! There are a multiple cases in stdlib that I found, I'll try to reduce.

slavapestov replied:
Thanks! I'm not sure about conditional requirements, but certainly adding constraints with generic parameters in them sounds like an oversight... we might be doing something wrong with those calls to getSuperclass().

xedin replied:
Indeed, I was surprised that we end up with dependent members there but I think it's because of getContextualSubstitutions since base of the dependent member type is a type variable. I'm going to finish something up first and come back to this.

slavapestov replied:
I see, if the base is a dependent member variable, I guess you'll still need your special operation for opening those up. The other alternative is that we can teach Type::subst() how to do that.

xedin replied:
Yes, that's what I was thinking actually - have a version of `getSupertype` and `conformsToProtocol` which accepts a callback that's going to property handle dependent member types, WDYT?

xedin committed:
[ConstraintSystem] Post-rebase fix to use new APIs for failing constraints

xedin committed:
[ConstraintSystem] Adjust anchors for new locators connected to associated type

This couldn't be done previously because locators could only be
anchored on expressions.

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
